{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edc198dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from data import ThermoMutDB\n",
    "from interfaces import ProteinPredictionTask, ProteinPredictionReturnType\n",
    "from models import ESM3Model\n",
    "from utils import ProteinComparator, ProteinComparatorMethod, ProteinAlignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af2bdd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instantiate components\n",
    "model = ESM3Model()\n",
    "comparator = ProteinComparator()\n",
    "test_set = ThermoMutDB(filepath=\"edited_thermomutdb_subset_esm3.json\", mutation_manifest_path=\"../data/thermomutdb_alphafold_investigation.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c4264e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUFFIX = \"mutant\" #@param ['mutant','unaltered'] {type:\"raw\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d55de5",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_set_results = []\n",
    "\n",
    "print(f\"Running through {len(test_set)} samples...\")\n",
    "\n",
    "# Run through all samples in the test set\n",
    "for idx, sample in enumerate(test_set):\n",
    "    try:\n",
    "        # Obtain necessary information\n",
    "        original_protein_fasta = sample[\"real_fasta\"].iloc[0][:-4]\n",
    "        mutation_code = sample[\"mutation_code\"].iloc[0]\n",
    "        target_mutation_id = sample[\"target_mutation_id\"].iloc[0]\n",
    "        real_protein_pdb = sample[\"real_pdb_wild\"].iloc[0]\n",
    "\n",
    "        # Capture the mutation index and amino acid change\n",
    "        original_amino_acid = mutation_code[0]\n",
    "        variant_position = int(mutation_code[1:-1]) - 1\n",
    "        variant_amino_acid = mutation_code[-1]\n",
    "\n",
    "        # If the variant is in the last 4 amino acids, we cannot predict for it with ESM3.\n",
    "        # Skip it.\n",
    "        if len(original_protein_fasta) <= variant_position:\n",
    "            continue\n",
    "        else:\n",
    "            # Make sure the amino acid in the FASTA sequence matches the original amino acid in the variant\n",
    "            sequence_position_matches_variant_original_amino_acid = original_amino_acid == original_protein_fasta[variant_position]\n",
    "\n",
    "            print(f\"Mutation ID: {target_mutation_id}\")\n",
    "            print(f\"Assert: {original_amino_acid} matches {original_protein_fasta[variant_position]} is {sequence_position_matches_variant_original_amino_acid}.\", end=\" \")\n",
    "            \n",
    "            # Filter if amino acid in the sequence does not match original amino acid in the variant\n",
    "            if not sequence_position_matches_variant_original_amino_acid:\n",
    "                print(\"Skipping...\")\n",
    "                continue\n",
    "            else:\n",
    "                print()\n",
    "\n",
    "        if SUFFIX == \"mutant\":\n",
    "            # Modify the sequence to be the variant sequence\n",
    "            original_protein_fasta = original_protein_fasta[:variant_position] + variant_amino_acid + original_protein_fasta[variant_position + 1:]\n",
    "\n",
    "        # Run inference on the original protein FASTA\n",
    "        resultant_protein = model(\n",
    "            ProteinPredictionTask.STRUCTURE_PREDICTION,\n",
    "            protein=original_protein_fasta,\n",
    "            generation_config_kwargs={ \"num_steps\": 1, \"temperature\": 0.0 },\n",
    "            return_type=ProteinPredictionReturnType.DEFAULT,\n",
    "        )\n",
    "        resultant_pdb = resultant_protein.to_pdb_string()\n",
    "\n",
    "        # Compute the optimal alignment and USalign/TMalign scores\n",
    "        resulting_alignment = comparator.compute_score_and_alignment(\n",
    "            resultant_pdb,\n",
    "            real_protein_pdb,\n",
    "        )\n",
    "\n",
    "        # Print and visualize the protein alignment and scores\n",
    "        for alignment_score in resulting_alignment:\n",
    "            print(f\"{alignment_score.method.value}: {alignment_score.final_score}\")\n",
    "        \n",
    "        # Append the current alignment to the results\n",
    "        test_set_results.append((idx, resulting_alignment + [\n",
    "            ProteinAlignment(\n",
    "                method=ProteinComparatorMethod.LDDT,\n",
    "                pdb1=resultant_pdb,\n",
    "                pdb2=None,\n",
    "                superimposed_pdb=None,\n",
    "                score1=resultant_protein.plddt,\n",
    "                score2=resultant_protein.plddt[variant_position].item(), # The confidence of the atomic coordinates of the amino-acid/residue at the variant position\n",
    "                final_score=np.average(resultant_protein.plddt),\n",
    "                auxiliary=None,\n",
    "            )\n",
    "        ]))\n",
    "\n",
    "        # Visualize the alignment\n",
    "        # comparator.visualize_alignment(resulting_alignment[0])\n",
    "    except Exception as e:\n",
    "        # Skip the sample due to the exception\n",
    "        print(f\"Skipping sample at index {idx} due to `{e}`.\")\n",
    "        test_set_results.append((idx, e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08c86e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(test_set_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36eb4d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXTRA_FIELDS = [\"target_id\", \"pdb_id\", \"mutation_code\"]\n",
    "\n",
    "data =  { field_name: [] for field_name in EXTRA_FIELDS }\n",
    "data.update({\n",
    "    \"US-Align\": [],\n",
    "    \"pLDDT_of_variant_residue\": [],\n",
    "    \"avg_pLDDT\": [],\n",
    "})\n",
    "\n",
    "for i, sample in enumerate(test_set):\n",
    "    results = list(filter(lambda res: i == res[0], test_set_results))\n",
    "    if len(results) > 0:\n",
    "        _, alignment_list_or_err = results[0]\n",
    "        if isinstance(alignment_list_or_err, Exception):\n",
    "            continue\n",
    "        us_alignment, plddt_alignment = tuple(alignment_list_or_err)\n",
    "        for field in EXTRA_FIELDS:\n",
    "            data[field] = data[field] + [sample[field].iloc[0]]\n",
    "\n",
    "        data[\"US-Align\"] = data[\"US-Align\"] + [us_alignment.final_score]\n",
    "        data[\"pLDDT_of_variant_residue\"] = data[\"pLDDT_of_variant_residue\"] + [plddt_alignment.score2]\n",
    "        data[\"avg_pLDDT\"] = data[\"avg_pLDDT\"] + [plddt_alignment.final_score]\n",
    "\n",
    "pd.DataFrame(data).to_csv(f\"thermomutdb_prediction_results_{SUFFIX}.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
